### ISSUE TITLE
`once_cell` Build Error

### ISSUE THREAD

**Issue by zedseven** [kangalioo/poise#35]
In updating my bot to the latest versions of dependencies, I encountered a build error with Poise:
```
error[E0433]: failed to resolve: could not find `sync` in `once_cell`
  --> C:\Users\Zacc\.cargo\git\checkouts\poise-2ca7e75d46548da7\e94eb68\src\framework\mod.rs:14:27
   |
14 |     user_data: once_cell::sync::OnceCell<U>,
   |                           ^^^^ could not find `sync` in `once_cell`

error[E0433]: failed to resolve: could not find `sync` in `once_cell`
  --> C:\Users\Zacc\.cargo\git\checkouts\poise-2ca7e75d46548da7\e94eb68\src\framework\mod.rs:76:41
   |
76 |             user_data: once_cell::sync::OnceCell::new(),
   |                                         ^^^^^^^^ not found in `once_cell::sync`
   |
help: consider importing one of these items
   |
6  | use core::lazy::OnceCell;
   |
6  | use once_cell::unsync::OnceCell;
   |
6  | use std::lazy::OnceCell;
   |
6  | use tokio::sync::OnceCell;
   |
```

I'm almost certain it's because of commit 9243918b8523a9ff5c126f4ff6d33903c6d30c9e, which removed the default features for dependencies and therefore removed the `std` feature in `once_cell`.

I've tested this in a blank new project and the `sync` module is not available without the `std` feature.

Thank you for your work on this framework.

**Comment by kangalioo** [comment]
Oh man, thanks for catching this

I thought the Cargo.toml change was fine because CI passed and my bots compiled, but it seems like it only worked because rustls has a transitive dependency on once_cell with std feature enabled. So only with a non-rustls backend does this occur

**Comment by zedseven** [comment]
No problem, thanks for fixing it quickly.

I was wondering why the CI checks passed - it had me concerned it was a problem on my end somewhere.

**Comment by kangalioo** [comment]
Yeah I'm currently working on testing native TLS in CI, so that this doesn't happen again. I'd like to wait with the fix until that is done to make sure the new CI can actually catch this build error.

It will take a way because yaml is awful and GHA is finnicky (you have to wait out the entire compilation process every time you make a change, including Rust toolchain + crates.io index + crates download...), I hope this is not too big of an issue for you

**Comment by zedseven** [comment]
Yeah, I know how painful CI config changes are; I went through a similar thing with Travis CI, so I think it's just a general problem.

It's not a huge issue by any means. I didn't update for a particular reason - it was just to keep things up to date - so I'll just hold off on updating until things get resolved.

Worst case, I'll fork the repo temporarily and use the fix there until it's implemented upstream, so there's no rush.

**Comment by kangalioo** [comment]
I do wonder how the build managed to break for you

The `openssl` crate, which is used with native TLS, unconditionally depends on once_cell with default features, i.e. including the std feature ([link](https://github.com/sfackler/rust-openssl/blob/master/openssl/Cargo.toml#L28)). That's probably why I cannot reproduce the build error in CI.

I think I will just fix the Cargo.toml and hope that your issue doesn't occur again, whatever it was caused by

**Comment by zedseven** [comment]
I don't see any mention of the `openssl` crate in my `cargo tree` output: [radium-cargo-tree.txt](https://github.com/kangalioo/poise/files/7735227/radium-cargo-tree.txt)

If you want to see the issue for yourself, my repo is public and you just have to run a `cargo update` in it to see the problem: https://github.com/zedseven/radium

(if you push a fix to the master branch of Poise, you might need to set the commit manually in the `Cargo.lock` to reproduce the problem)

**Comment by kangalioo** [comment]
Ohh, I was mistaken and your project does actually use rustls. But rustls's transitive dependency on once_cell is only active on Unix-like systems, so I assume you're building on Windows.

Case closed then, thanks for debugging with me :smile:

Issue should be fixed in 8194e079e2189c5970943908b48728a974c0b9d8 btw

**Comment by zedseven** [comment]
Ah, that makes sense. Yes, unfortunately for now my dev environment is still Windows.

No problem! Thank you for being so quick to fix it and thank you again for making such a great bot framework.
